Abstract Ops for batch normalization

gvtulder opened the pull request:
This is a project to define abstract batch normalization Ops that can get optimized to use the cuDNN Ops, while providing an alternative pure Theano implementation that can run on CPUs and on GPUs if cuDNN is not available.

I've started by copying the interface from the gpuarray backend, but this might also be an opportunity to improve some things.

See also issue #5187.

gvtulder commented:
To summarize: this introduces two functions and three Ops that closely mirror the structure of the cuDNN Ops.

- `theano.tensor.bn.batch_normalization_train`
- `theano.tensor.bn.batch_normalization_test`
- `theano.tensor.bn.AbstractBatchNorm`
- `theano.tensor.bn.AbstractBatchNormInference`
- `theano.tensor.bn.AbstractBatchNormGrad`

The Ops have a Python implementation. There are optimizations that replace them with Theano Ops. Additional optimizations should introduce the cuDNN-based Ops if they are available.

Would it be useful to have a separate `AbstractBatchNormInferenceGrad` Op? At this point, the gradient of the inference Op is implemented in Theano, similar to how it is done in the cuDNN equivalent.

Using variance everywhere instead of `stdinv` is probably more consistent. I am not sure if this would introduce problems with the precision. (The batch normalization code is a bit sensitive anyway.)

Moving averages and in-place flags can also be added.

f0k commented on theano/tensor/nnet/bn.py:
```diff
@@ -73,3 +79,387 @@ def batch_normalization(inputs, gamma, beta, mean, std,
+
+    Notes
+    -----
+    Requires cuDNN 5 and Theano 0.9dev2 or more recent.
```
This sentence can be removed.

gvtulder replied:
Yes, although it might be useful to mention the cuDNN optimization somewhere in the documentation.

f0k commented on theano/tensor/nnet/bn.py:
```diff
@@ -73,3 +79,387 @@ def batch_normalization(inputs, gamma, beta, mean, std,
+                         "as inputs; got %d and %d instead of %d" %
+                         (gamma.ndim, beta.ndim, ndim))
+    if epsilon < 1e-5:
+        raise ValueError("epsilon must be at least 1e-5, got %f" % epsilon)
```
Not sure what to do about this -- this is a strict requirement for cuDNN, but not for the Theano code (although setting it too small may not work well). I'd suggest to remove this, what do others think?

gvtulder replied:
Checking epsilon here makes the move to cuDNN a bit more predictable, because the optimization will never fail because of an epsilon that is too small. If epsilon is only checked in the optimization, it's more difficult for the user to find out why it doesn't use cuDNN.
If this check is removed here, the minimum value for cuDNN should be documented, or maybe even produce a warning. Is there a good reason to set epsilon lower than the cuDNN limit? If there isn't, I think it's better to check this here.

f0k commented on theano/tensor/nnet/bn.py:
```diff
@@ -73,3 +79,387 @@ def batch_normalization(inputs, gamma, beta, mean, std,
+    return batchnorm_op(inputs, gamma, beta, mean, var, epsilon=epsilon)
+
+
+class AbstractBatchNorm(Op):
```
Maybe this should be called `AbstractBatchNormTrain` instead.

f0k commented on theano/tensor/nnet/bn.py:
```diff
@@ -73,3 +79,387 @@ def batch_normalization(inputs, gamma, beta, mean, std,
+    ----------
+    mode : {'per-activation', 'spatial'}
+        Whether to normalize per activation or share normalization factors
+        across spatial dimensions (i.e., all dimensions past the second).
```
I'd prefer to have an `axes` argument here, to freely specify the axes to be normalized over. Not all cases can be replaced with the cuDNN implementation then, but we shouldn't restrict the interface because of that.

gvtulder replied:
An `axes` argument would be nice, although I do find the two modes convenient. Perhaps the `axes` argument could have special values `'per-activation'` and `'spatial'` for these common cases?

f0k commented on theano/tensor/nnet/bn.py:
```diff
@@ -73,3 +79,387 @@ def batch_normalization(inputs, gamma, beta, mean, std,
+
+        mean = x.mean(axes, keepdims=True)
+        stdinv = 1.0 / numpy.sqrt(x.var(axes, keepdims=True) + epsilon)
+        out = (x - mean) * scale * stdinv + bias
```
This should have `(scale * stdinv)` in brackets for improved performance (multiply the smaller tensors first).

f0k commented on theano/tensor/nnet/bn.py:
```diff
@@ -73,3 +79,387 @@ def batch_normalization(inputs, gamma, beta, mean, std,
+
+    def perform(self, node, inputs, output_storage):
+        x, scale, bias, estimated_mean, estimated_variance, epsilon = inputs
+        out = (x - estimated_mean) * scale / numpy.sqrt(estimated_variance + epsilon) + bias
```
Again, this should multiply the smaller tensors first.

f0k commented on theano/tensor/nnet/bn.py:
```diff
@@ -73,3 +79,387 @@ def batch_normalization(inputs, gamma, beta, mean, std,
+
+    mean = x.mean(axes, keepdims=True)
+    stdinv = T.inv(T.sqrt(x.var(axes, keepdims=True) + epsilon))
+    out = (x - mean) * scale * stdinv + bias
```
`(scale * stdinv)`

f0k commented on theano/tensor/nnet/bn.py:
```diff
@@ -73,3 +79,387 @@ def batch_normalization(inputs, gamma, beta, mean, std,
+        return None
+
+    # TODO copy_stack_trace?
+    return [(x - estimated_mean) * scale / numpy.sqrt(estimated_variance + epsilon) + bias]
```
`(scale / T.sqrt(...))` (i.e., put in brackets, and use `T` instead of `numpy`)

gvtulder commented:
I've added some tests (based on those for gpuarray).

gvtulder commented:
The last two commits add in-place options to the gpuarray Opts. The in-place updates for the running_mean and running_var seems to work (commit a3d8ec3).

The in-place updates for the input/output of batch_normalization_train (919f569) break the tests, for some reason. I do not know why this happens. The gradient computation no longer works if the in-place option is enabled. I have the impression that this might be because the GpuDnnBatchNormGrad depends on the original input, but that this original input is changed by GpuDnnBatchNorm before the gradient is computed.

gvtulder commented:
Found a temporary solution to the inplace problem by renaming the property. (See #5220.) The Jenkins GPU tests now pass.

f0k commented:
Sorry, I was busy. Wow, you're making some good progress there!

> The gradient computation no longer works if the in-place option is enabled. I have the impression that this might be because the GpuDnnBatchNormGrad depends on the original input, but that this original input is changed by GpuDnnBatchNorm before the gradient is computed.

Hmm, I think the inplace optimization shouldn't be applied if other clients need the input. But I'm not familiar with how the inplace optimization works in Theano, your registration seems proper. We'll need to summit some of the Theano devs for this. @nouiz?

Nice thought for `ndim > 5`, but this only works for `mode == 'spatial'`! Otherwise you're mixing up dimensions that should be normalized independently.

Generally your code looks very good. I can't do a thorough review now. Let's see what the devs have to say about the inplace problem. Good job so far!

gvtulder commented:
@f0k: The inplace problem is solved for the time being, but it's not really a good solution because it now works only because I reordered the optimizations. (See #5223.) Perhaps the optimization code isn't really expecting Ops to have multiple inplace optimizations that change the `destroy_map`.

> Nice thought for `ndim > 5`, but this only works for `mode == 'spatial'`! Otherwise you're mixing up dimensions that should be normalized independently.

Are you sure? I think it works for `mode == 'per-activation'` as well because that's only computing the averages across the 0 dimension. All the other activations are normalized independently, so there the number of dimensions shouldn't be a problem. I was a bit confused about this at first because cuDNN is so strict about 4D and 5D, but now I think it should work. (And I think it's tested as well.)

I'm not entirely sure how this works for non-contiguous inputs, though. That might in theory mess up the reshape and unreshape steps, but I assume the reshape Op handles non-contiguous data correctly.

f0k commented:
> The inplace problem is solved for the time being [...] Perhaps the optimization code isn't really expecting Ops to have multiple inplace optimizations

I just saw #5223. I guess you're right -- there were no Ops with independent inplace arguments until now, so you've uncovered a bug in the inner workings of Theano.

> Are you sure? I think it works for mode == 'per-activation' as well because that's only computing the averages across the 0 dimension.

It can be confusing, but yes: If you take a 6D numpy array `X` and compute `X.mean(axis=0)`, you get a 5D numpy array. If you reduce `X` to 5 dimensions before taking the mean across axis 0, it will average the last two dimensions together, which is not what we want. Your current code should actually fail with mismatched dimensions for 6D input, but you didn't extend the tests to check this.

gvtulder commented:
I'm not entirely sure what you mean with 'mismatched dimensions', but this seems to work:
```python
import numpy as np

shape = (2, 5, 7, 9)
x = np.array(range(0, np.prod(shape))).reshape(shape)
x_mean = x.mean(axis=0, keepdims=True)

x_reduced = x.reshape(shape[:2] + (-1,))
x_reduced_mean = x_reduced.mean(axis=0, keepdims=True)

x_mean_restored = x_reduced_mean.reshape((1,) + shape[1:])

print 'x_mean.shape', x_mean.shape  # (1, 5, 7, 9)
print 'x_mean_reduced.shape', x_mean_restored.shape  # (1, 5, 7, 9)
print np.all(x_mean == x_mean_restored)  # True
```
I'll have to think about this, because I might be missing something important.

f0k commented:
> I'll have to think about this, because I might be missing something important.

Oh, boy... you're right. With `axis=0`, it will collapse the first dimension and all the others can still be reshaped at will. I should go home and get some rest!

> I was a bit confused about this at first because cuDNN is so strict about 4D and 5D

Apparently that's because they don't want to support more strides. We can leave that up to Theano (it will copy the tensor if it can't be reshaped by just taking a different view at the memory).

> (And I think it's tested as well.)

Indeed, it is: https://github.com/Theano/Theano/pull/5190/files#diff-62bff40f5ac85ec667d250a8198c5557L1364
I thought I had seen it without tensor6. Again, it's time to go home! :)

nouiz commented:
We try to make an release candidate next week. I don't think we will have the times to review this well before. I think it would be great to include that in a second release candidate. This would make the release don't have GPU ops in the user graph (except for cudnn rnn). And that would be great to have.

gvtulder commented:
Rebased to fix a merge conflict.
The abstract helper methods `theano.tensor.nnet.bn.batch_normalization_train` and  `theano.tensor.nnet.bn.batch_normalization_test` now also accept non-broadcasted parameters. For spatial normalization, for example, you can now provide a `vector` scale instead of a `tensor4` with three 1-dimensions.

f0k commented:
> The abstract helper methods `theano.tensor.nnet.bn.batch_normalization_train` and `theano.tensor.nnet.bn.batch_normalization_test` now also accept non-broadcasted parameters.

I think that's a good idea. It complicates the code somewhat, but is confined to the helper functions.

Just to understand the relation to #5223: The code currently works because of the order you register the inplace optimizations? So #5223 is not strictly needed for this PR, but should be merged to prevent problems in the future?

gvtulder commented:
I don't know. The test works because the inplace optimizations run in a lucky order, but I don't know if this is always the case. For example, I don't know if the optimizations are always run only once. If there is a possibility that they are run more than once that would lead to problems, because the second call will succeed where the first call failed.

f0k commented on theano/tensor/nnet/bn.py:
```diff
@@ -209,8 +209,12 @@ def batch_normalization_train(inputs, gamma, beta, axes='per-activation',
                          (running_var.ndim, params_ndim))
+
+    # epsilon will be converted to floatX later. we need to check
+    # for rounding errors now, since numpy.float32(1e-5) < 1e-5.
```
Oh. Good catch!

gvtulder commented:
Jenkins failed the gpuarray tests. It looks like this is because the libgpuarray version changed from 0.1 to 1.0. Is Jenkins reusing the Theano build cache from previous runs?

slefrancois commented:
Indeed it was stuck with an old cache. jenkins test this

slefrancois commented:
jenkins test this

nouiz commented:
jenkins failure are not related to this PR and have been fixed in master.

nouiz commented on theano/tensor/nnet/bn.py:
```diff
@@ -75,3 +81,625 @@ def batch_normalization(inputs, gamma, beta, mean, std,
+        if new_running_mean.broadcastable != running_mean.broadcastable:
+            new_running_mean = T.patternbroadcast(new_running_mean, running_mean.broadcastable)
+        if new_running_var.broadcastable != running_var.broadcastable:
+            new_running_var = T.patternbroadcast(new_running_var, running_var.broadcastable)
```
I think those 2 if aren't needed, but you can let them.

nouiz commented on theano/tensor/nnet/tests/test_bn.py:
```diff
@@ -122,21 +123,293 @@ def bn_ref(x, G, B, M, V):
+            f = theano.function([x, scale, bias, running_mean, running_var, dy],
+                                [out, x_mean, x_invstd, out_running_mean, out_running_var,
+                                 out2, x_mean2, x_invstd2, out_running_mean2, out_running_var2] +
+                                grads + grads2, mode='FAST_RUN')
```
Why do you put mode here? I think it is better to let the default mode. This will allow more testing, as travis also test FAST_COMPILE and the daily buildbot run with DebugMode. It would be great to have them tested too.

gvtulder replied:
I don't think I had a specific reason for that, it probably was copied over from somewhere else. I've removed it here and in the other places you mentioned.

nouiz commented on theano/tensor/nnet/tests/test_bn.py:
```diff
@@ -122,21 +123,293 @@ def bn_ref(x, G, B, M, V):
+    # backward pass
+    grads = T.grad(None, wrt=[x, scale, bias], known_grads={out: dy})
+    # compile
+    f = theano.function([x, scale, bias, dy], [out, x_mean, x_invstd] + grads, mode='FAST_RUN')
```
idem, remove the mode parameter.

nouiz commented on theano/tensor/nnet/tests/test_bn.py:
```diff
@@ -122,21 +123,293 @@ def bn_ref(x, G, B, M, V):
+
+            # compile to compute all differences
+            f = theano.function([x, scale, bias, running_mean, running_var],
+                                T.sum(sum(results)), mode='FAST_RUN')
```
idem

nouiz commented on theano/tensor/nnet/tests/test_bn.py:
```diff
@@ -122,21 +123,293 @@ def bn_ref(x, G, B, M, V):
+            grads2 = T.grad(None, wrt=[x, scale, bias, mean, var], known_grads={out2: dy})
+            # compile
+            f = theano.function([x, scale, bias, mean, var, dy],
+                                [out, out2] + grads + grads2, mode='FAST_RUN')
```
idem

nouiz commented on theano/tensor/nnet/tests/test_bn.py:
```diff
@@ -122,21 +123,293 @@ def bn_ref(x, G, B, M, V):
+    # compile
+    f = theano.function([x, scale, bias, mean, var, dy],
+                        [out_train, x_mean, x_invstd, out_test] + grads_train + grads_test,
+                        mode='FAST_RUN')
```
idem.

nouiz commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -1379,87 +1381,253 @@ def test_dnn_batchnorm_train():
+                utt.assert_allclose(outputs_gpu[2], outputs_ref[2])  # invstd
+                utt.assert_allclose(outputs_gpu[3], outputs_ref[3])  # running_mean
+                utt.assert_allclose(numpy.nan_to_num(outputs_gpu[4]),
+                                    numpy.nan_to_num(outputs_ref[4]))  # running_var
```
I didn't know about nan_to_num. Good solution. Ideally, we should find inputs range that won't give that, but I think this is a good enough solution. No need to change this.

nouiz commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -1379,87 +1381,253 @@ def test_dnn_batchnorm_train():
+                                         mode=mode_with_gpu)
+            f_ref = theano.function([x, scale, bias, running_mean, running_var, dy],
+                                    [out_ref, x_mean_ref, x_invstd_ref,
+                                     out_running_mean_ref, out_running_var_ref] + grads_ref)
```
Pass `mode=mode_without_gpu`, to make sure the GPU isn't activated. It help make some tests pass when device=cuda, even if we don't officialy support that.

nouiz commented:
I have a local fix for the failure in FAST_COMPILE. I'll push this to this branch with the change to the opt.

nouiz commented:
I don't have the right to push to your branch. Can you enable it?

I have updated the optimization to the new way in the new back-end and did the small fixes I commented about:

https://github.com/nouiz/Theano/tree/gvtulder-f-batchnorm-abstract

Why did you add the automatic flatten in the gpu code?

thanks

gvtulder commented on theano/gpuarray/opt.py:
```diff
@@ -2005,3 +2005,28 @@ def local_gpu_solve(op, context_name, inputs, outputs):
+# Register cuDNN batch normalization implementation
+abstract_batch_norm_groupopt = theano.gof.optdb.LocalGroupDB()
+abstract_batch_norm_groupopt.__name__ = "gpuarray_batchnorm_opts"
+register_opt('fast_compile')(abstract_batch_norm_groupopt)
```
No specific reason for the LocalGroupDB, other than that this was how it is done for the convolution ops and I'm still trying to understand the different ways to register optimizations.

gvtulder commented:
Thanks. I've added your commits (you should now also be able to push to this branch yourself).

I've added `__setstate__` methods to the ops with new properties. Would this need tests? (The other `__setstate__` methods in this file do not seem to have them.)

I might be misunderstanding your objections to the flattening, because I think it should work fine.

The idea behind the flattening in the GPU code is to allow you to use the cuDNN functions on inputs with more than five dimensions. There is an "unflattening" at the [end of the function](https://github.com/gvtulder/Theano/blob/b690578d54cd923323d97238352d8dfeefbc860e/theano/gpuarray/dnn.py#L2614-L2616) that restores everything back to its original shape, so I would expect it to work fine. There is no flattening and unflattening for the CPU/pure Theano version because it's not required there.

In the GPU functions, the flatten/unflatten for >5D is quite similar to the padding-to-four-dimensions that is already being done for inputs with less than four dimensions, so it seemed a logical thing to add. Does that make sense or did I completely misunderstand your objections? When could this be a problem?

Alternatively, this transformation could be moved to the optimizations. It would be a bit strange to handle <4D and >5D in different places, though.

nouiz commented:
About the flatten, I don't understand what the 6d dimensions mean. Can you explain? Is there only 1 interpretation?

Can you confirm that calling batch_normalization_train or  dnn_batch_normalization_train on a 6d tensor will give the same result for both mode? At first, I think it wasn't the case, now I think it is the case. If it is the case and it is tested (I think so from review), I'm good with the flatten.

nouiz commented:
LocalGroupDB are useful when we want a group of optimization to be executed
in a specific order inside an EquilibriumOptimizer.


Here, there is only 1 conversion possible per op as there is only 1 GPU
implementation. On the CPU, it is the same. As the CPU optimization are
already always after the GPU optimization, we don't need to force that
order.

Also, the way it is currently implemented, if we can't run it on the GPU as
if cudnn isn't there (or opencl is enabled... not ready) the CPU
optimization will get applied but we will have done this too late, so it
won't run on the GPU. So I need to revisit this and this would request an
LocalGroupDB!

I'll try to do this today.

On Tue, Jan 31, 2017 at 6:17 AM, Gijs van Tulder <notifications@github.com>
wrote:

> *@gvtulder* commented on this pull request.
> ------------------------------
>
> In theano/gpuarray/opt.py <https://github.com/Theano/Theano/pull/5190>:
>
> > @@ -2005,3 +2005,28 @@ def local_gpu_solve(op, context_name, inputs, outputs):
>                                 local_abstractconv3d_gradinputs_gemm, 30,
>                                 'conv_gemm',
>                                 'gpuarray', 'fast_compile', 'fast_run')
> +
> +
> +# Register cuDNN batch normalization implementation
> +abstract_batch_norm_groupopt = theano.gof.optdb.LocalGroupDB()
> +abstract_batch_norm_groupopt.__name__ = "gpuarray_batchnorm_opts"
> +register_opt('fast_compile')(abstract_batch_norm_groupopt)
>
> No specific reason for the LocalGroupDB, other than that this was how it
> is done for the convolution ops and I'm still trying to understand the
> different ways to register optimizations.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/5190>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-zEPfGcSZ5P-qgXpvp7H3L2-qoESks5rXxhmgaJpZM4KrV74>
> .
>

gvtulder commented:
Yes, there are tests for 6D tensors in the gpu tests. They compare the results of `dnn_batch_normalization_train`, the  abstract `batch_normalization_train` and a reference implementation that doesn't do any flattening. In practice it seems to work.

It should also work on paper:
- With `per-activation`, the average is removing axis 0. With 6 dimensions, the input gets flattened to `[0, 1, 2, 3, 4 + 5]`. The mean taken over 0 will be `[x, 1, 2, 3, 4 + 5]`, which is reshaped to `[x, 1, 2, 3, 4, 5]` at the end.
- With `spatial`, the average is removing all axes except 1. With 6 dimensions, it should remove axes 0, 2, 3, 4, 5. The input is first flattened to `[0, 1, 2, 3, 4 + 5]`. The mean taken over 0, 2, 3, 4 will be `[x, 1, x, x, x + x]`, which is finally reshaped back to `[x, 1, x, x, x, x]`.

As long as you don't mix averaged and non-averaged axes, I think it works fine.

nouiz commented:
I think all is good, so merging.

thanks!

nouiz commented:
I merged this PR too quickly, here is another PR with follow up:

https://github.com/Theano/Theano/pull/5474

gvtulder commented:
Good. Thanks for reviewing and merging.

f0k commented:
:clap: to @gvtulder, very good work!

nouiz commented on theano/tensor/nnet/bn.py:
```diff
@@ -75,3 +82,631 @@ def batch_normalization(inputs, gamma, beta, mean, std,
+        return [shape[0], shape[2], shape[2]]
+
+    def perform(self, node, inputs, output_storage):
+        x, dy, scale, x_mean, x_invstd, epsilon = inputs
```
epsilon isn't used in the perform nor in the opt that convert this to a graph. Is this normal? CuDNN do request it.

If it is normal that it isn't used and the only reason to add it here is that cudnn request it, then I'll add a comment about that.

f0k replied:
It's normal. In cudnnBatchNormalizationBackward, savedMean and savedInvVariance are optional. If they're given, epsilon shouldn't be needed, but it's required by cudnnBatchNormalizationBackward anyway and they scare you by saying it must be the same than in the forward pass.
(We could make `x_mean` and `x_invstd` optional in `AbstractBatchNormTrainGrad` as well, but there's no strong use case. It would trade a little bit of memory against computation time.)
